src/Disks/ObjectStorages/InMemoryDirectoryTree.cpp

```diff
@@ -0,0 +1,343 @@
+#include <Disks/ObjectStorages/InMemoryDirectoryTree.h>
+
+#include <Common/CurrentMetrics.h>
+#include <Common/Exception.h>
+
+#include <deque>
+#include <mutex>
+#include <filesystem>
+#include <memory>
+#include <ranges>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+static std::filesystem::path normalizePath(std::string_view path)
+{
+    if (path.ends_with('/'))
+        path.remove_suffix(1);
+
+    if (path.starts_with('/'))
+        path.remove_prefix(1);
+
+    return std::filesystem::path(path);
+}
+
+struct InMemoryDirectoryTree::INode
+{
+    std::weak_ptr<INode> parent;
+    std::string last_directory_name;
+
+    std::optional<DirectoryRemoteInfo> remote_info = {};
+    std::unordered_map<std::string, std::shared_ptr<INode>> subdirectories = {};
+
+    explicit INode(std::shared_ptr<INode> parent_, std::string last_directory_name_)
+        : parent(parent_)
+        , last_directory_name(std::move(last_directory_name_))
+    {
+    }
+
+    bool isVirtual() const
+    {
+        return !remote_info.has_value();
+    }
+
+    bool isRoot() const
+    {
+        return parent.lock() == nullptr;
+    }
+};
+
+std::shared_ptr<InMemoryDirectoryTree::INode> InMemoryDirectoryTree::walk(const std::filesystem::path & path, bool create_missing) const
+{
+    chassert(normalizePath(path.string()) == path);
+    std::shared_ptr<INode> node = root;
+
+    for (const auto & step : fs::path(path))
+    {
+        if (auto it = node->subdirectories.find(step); it != node->subdirectories.end())
+            node = it->second;
+        else if (create_missing)
+            node = node->subdirectories.emplace(step, std::make_shared<INode>(node, step)).first->second;
+        else
+            return nullptr;
+    }
+
+    return node;
+}
+
+void InMemoryDirectoryTree::traverseSubtree(const std::filesystem::path & path, std::function<void(const std::string &, const std::shared_ptr<INode> &)> observe) const
+{
+    chassert(normalizePath(path.string()) == path);
+    std::deque<std::pair<fs::path, std::shared_ptr<INode>>> unvisited;
+
+    if (auto start_node = walk(path))
+        unvisited.emplace_back("", std::move(start_node));
+
+    while (!unvisited.empty())
+    {
+        auto [node_path, node] = std::move(unvisited.front());
+        unvisited.pop_front();
+
+        observe(node_path, node);
+
```

Comment:
unvisited.reserve(unvisited.size() + node->subdirectories.size())

Reply:
changed to dfs